service/src/main/java/bio/terra/workspace/service/resource/controlled/ControlledResourceService.java

```diff
@@ -112,7 +118,21 @@ public String createAzureRelayNamespace(
                 resource, privateResourceIamRole, jobControl, resultPath, userRequest)
             .addParameter(ControlledResourceKeys.CREATION_PARAMETERS, creationParameters);
 
-    String jobId = jobBuilder.submit();
+    String jobId =
+        jobBuilder.submitWithCallback(
+            ControlledAzureRelayNamespaceResource.class,
+            new FutureCallback<>() {
+              @Override
+              public void onSuccess(
+                  JobResultOrException<ControlledAzureRelayNamespaceResource> result) {
+                if (result.getException() == null) {
```

Comment:
Instead of modifying all these Service classes, I think it's cleaner to just modify WorkspaceDao.java. In createWorkspace/updateWorkspace, you call activityLogDao.setWorkspaceUpdateDate().

[Here's an example ](https://github.com/DataBiosphere/jade-data-repo/blob/8a36c374c4b4a50f8851534d07c2ac1d5f818da8/src/main/java/bio/terra/service/filedata/azure/tables/TableDao.java#L68)of a Dao injecting another Dao.

The service approach is fragile. If someone adds a new service, they might forget to set workspace update date.

@ddietterich thoughts?

Reply:
ah i was not sure if Dao calling Dao was an anti-pattern.

But one issue with doing it in DAO is that it's more noisy especially in the case of stairway undo. E.g. during resource creation, we first store metadata in the resource db, and then create a gcp resources. If it fails, we delete the resource in the db. If we log in the resourceDao, we will have two updates in this failure flight but the end result is that nothing is updated.  https://github.com/DataBiosphere/terra-workspace-manager/blob/6f8d928494244f5464f7c588b87429080fe9d923/service/src/main/java/bio/terra/workspace/service/resource/controlled/flight/create/StoreMetadataStep.java#L12

Reply:
The problem with Dao calling Dao is making sure transactions are handled properly. Spring injects a transaction layer around every public call with an `@*Transaction` annotation. I feel strongly the code should not nest methods with those annotations. I know that it works, but it has led to confusion in reviewing and debugging code, so I think it is best avoided.

Is it an option to make a "Dao" that just had methods on the activity table, but those were always called from a "real" Dao with the `@*Transaction` marks?